File: Problems_and_Solutions/days-10 to 19/day-18/c++/solution 2.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n=prices.size();
        int min=prices[0];
        int i=0;
        int maxProfit=0;
        int temp=0;
        for(;i<n-1;i++)
        {
            if(prices[i]<prices[i+1])
            {
                if(prices[i]<min)
                    min=prices[i];
                temp=prices[i+1]-min;
                if(temp>maxProfit)
                    maxProfit=temp;
            }
        }
        return maxProfit;
    }
};

File: Problems_and_Solutions/days-10 to 19/day-18/Java/Solution2.java
class Solution2 {
    public static int maxProfit1(int[] prices) {
        int profit = 0;
        for(int i=0; i<prices.length; i++){
            for(int j=i+1; j<prices.length; j++){
                profit = Math.max(profit, prices[j] - prices[i]);
            }
        }
        return profit;
    }

    public static int maxProfit2(int[] prices) {
        int min = Integer.MAX_VALUE;
        int profit = 0;

        for(int price : prices){
            if(price < min){
                min = price;
            }else{
                profit = Math.max(profit, price-min);
            }
        }

        return profit;
    }

    public static void main(String args[]){
        int[] prices = {7, 1, 5, 3, 6, 4};
        System.out.println(maxProfit1(prices));
    }
}